### DISCUSSION TITLE
Allow custom names for branched targets

### DISCUSSION THREAD

**Discussion by JohannesNE** [ropensci/targets#668] [Ideas]
It would be very convenient to allow custom names of branched targets. Example:

I have 3 files in `data/`: `data/churn.csv` `data/churn2500.csv` and `data/churn5000.csv`

My pipeline looks like this:
```
list(
  tar_files(churn_file, list.files("data", full.names = TRUE)),
  tar_target(churn_data, read.csv(churn_file), map(churn_file), iteration = "list"),
  tar_target(churn_rows, nrow(churn_data), map(churn_data), iteration = "list")
)
```

```
names(tar_read(churn_data))
# [1] "churn_data_123e0f60" "churn_data_b5c463de"
# [3] "churn_data_f4ac93c0"
```

It would be very useful to assign names to `churn_data` branches based on `churn_file`, so I know which file each branch comes from. Also `churn_rows` could just inherit names from `churn_data`

The ideal behaviour would be something like this:

```
list(
  tar_files(churn_file, list.files("data", full.names = TRUE)),
  tar_target(churn_data, read.csv(churn_file), map(churn_file), iteration = "list", name = basename(churn_file)),
  tar_target(churn_nrow, nrow(churn_data), map(churn_data), iteration = "list", inherit_name = TRUE)
)
```

```
names(tar_read(churn_data))
# [1] "churn.csv"     "churn2500.csv"
# [3] "churn5000.csv"
```

```
names(tar_read(churn_nrow))
# [1] "churn.csv"     "churn2500.csv"
# [3] "churn5000.csv"
```

**Comment by wlandau** [comment]
This gets tricky because in other situations, dynamic branches are entire row groups of a data frame, which cannot be turned into names. In your case, each branch of `churn_nrow` could return a `tibble` with the number of rows and the name of the file. Then, `tar_read(churn_nrow)` would read all the branches and tell you the number of rows in each file.

**Reply by JohannesNE** [reply]
I see. 

For my (actual) project I have a number of computationally expensive steps calculating features for individual subjects. In the end, these are combined into a data.frame.

I want to be able to look up a branch corresponding to a specific subject. Ideally like `tar_load(feature1); feature1$subject2`.

If branches always come in the same order, I could just change the names when I load the target. Otherwise I have considered adding a custom name attribute to each branch and then write a function that loads a target and adds this attribute as names. Would this be reasonable or is there a better way to keep track of the branches?

**Reply by wlandau** [reply]
`tar_meta(your_target)$children[[1]]` lists the branch names in the order of the target you branched over. Otherwise, it is good practice to append metadata (e.g. a new column) identifying the characteristics of the current branch.

**Reply by wlandau** [reply]
Another technique is to define a new dynamic target that maps over the one you want to know about and returns something like `tibble(name = tar_name(), metadata = list(subject_IDs))`. Then this new downstream target would allow you to look up the upstream ones.

**Comment by JohannesNE** [comment]
My solution for now is to add an id attribute. In functions used by branched targets I'll return data using this function:

```
# Check that all upstream targets (...) have same id and pass that id on.
return_with_id <- function(obj, ...) {
  ids <- map_chr(list(...), attr, "id")
  assert_that(length(unique(ids))==1)
  id <- ids[1]
  
  structure(obj,
            id = id)
}
```

`obj` is the object to return. All upstream targets are passed to `...`. This way I both check that all upstream targets have the same id and ensure that the downstream target inherits this id.

**Reply by wlandau** [reply]
That will work. Another choice is to return a `tibble` with `id` in the column. There are situations where attributes are dropped, e.g. `dplyr::bind_rows()` or if you choose `format = "fst"` in `tar_target()`.

**Reply by JohannesNE** [reply]
That might be better. I end up binding it into a nested tibble anyway. I could then write a wrapper around tar_load(/read) that loads it as a list with names from the id column.

**Comment by JohannesNE** [comment]
I still think it would be very useful if branches had sensible names. It would make branching a lot more accessible to a lot of people. The random names are a bit uncomfortable (at least for me).

For `pattern = map(x, y)` the names could be `names(x)` 
For `pattern = cross(x, y)` the names could be `interaction(names(x), names(y), sep = "_")`
For `tar_group_by(data, produce_data(), var1, var2)` the names could be `"{var1}_{var2}"`

If the names are set in the pattern function e.g. `pattern = map(x, y, .names = names(x))`, the option could just not be there in the situations where it does not make sense (e.g. `with tar_group_size()`)

**Reply by JohannesNE** [reply]
NB. I will not pretend to know how to (or even if it is possible to) implement this in {targets}, but the naming of branches was a major hurdle for me when I was thinking about converting my script based analysis into a {targets}-based. After having committed to {targets}, I'm happy I did, regardless of the branching names.

**Comment by jdtrat** [comment]
Hi! I recently began using targets and love it! Thanks for such an amazing package, @wlandau! I'm looking into static branching and was wondering if there's a way to use a glue type syntax when defining the target value names? As I understand it, there's only a way to have the names argument from `tar_map()` act as suffixes as shown below:

```r
tar_map(
  values = tibble::tibble(
    data_set = rlang::syms(c("iris", "mtcars")),
    column = c("Petal.Length", "mpg"),
    sum_fun = rlang::syms(c("mean", "median")),
    suffix = c("petal_length", "mpg")
  ),
  tar_target(summarized,
             sum_fun(data_set[[column]])
    ),
  names = "suffix"
)
tar_manifest()
#> # A tibble: 2 × 3
#>   name                    command                          pattern
#>   <chr>                   <chr>                            <chr>  
#> 1 summarized_petal_length "mean(iris[[\"Petal.Length\"]])" <NA>   
#> 2 summarized_mpg          "median(mtcars[[\"mpg\"]])"      <NA>
```

I think it would be great to allow flexibility in target names for static branching as follows:

```r
tar_map(
  values = tibble::tibble(
    data_set = rlang::syms(c("iris", "mtcars")),
    column = c("Petal.Length", "mpg"),
    sum_fun = rlang::syms(c("mean", "median")),
  ),
  tar_target("{sum_fun}_{column}",
             sum_fun(data_set[[column]])
    )
)
# Desired Output of `tar_manifest()`:
tar_manifest()
#> # A tibble: 2 × 3
#>   name                    command                          pattern
#>   <chr>                   <chr>                            <chr>  
#> 1 mean_petal_length "mean(iris[[\"Petal.Length\"]])" <NA>   
#> 2 median_mpg          "median(mtcars[[\"mpg\"]])"      <NA>
```

I saw you mentioned the complexity of implementing custom names with cross (https://github.com/ropensci/targets/issues/672) but am not sure if the same issue would present itself here, so I thought I'd bring it up!

**Reply by wlandau** [reply]
Thanks, @JohannesNE. It's a neat thought, but I think you get most of that flexibility in other ways that already exist. In `tar_map()`, you have complete control over the suffix in the example you described. For complete control over the entire name, you can use `tar_eval()`: https://books.ropensci.org/targets/static.html#metaprogramming.

**Reply by jdtrat** [reply]
🤯That's amazing! Thank you!

``` r
tar_eval(tar_target(name,
                    sum_fun(data_set[[column]])),
         values = tibble::tibble(
           data_set = rlang::syms(c("iris", "mtcars")),
           column = c("Petal.Length", "mpg"),
           sum_fun = rlang::syms(c("mean", "median")),
           suffix = c("petal_length", "mpg"),
           name = rlang::syms(glue::glue("{sum_fun}_{suffix}"))
           )
)
tar_manifest()
#> # A tibble: 2 × 3
#>   name              command                          pattern
#>   <chr>             <chr>                            <chr>  
#> 1 mean_petal_length "mean(iris[[\"Petal.Length\"]])" <NA>   
#> 2 median_mpg        "median(mtcars[[\"mpg\"]])"      <NA>

**Reply by JohannesNE** [reply]
Oh... I should just use static branching!

**Reply by JohannesNE** [reply]
Though my graph is now less useful 
![image](https://user-images.githubusercontent.com/11048760/139072691-743f7772-4527-4bfa-ad1e-98fce2e2dc27.png)

**Reply by wlandau** [reply]
You could temporarily downsize `values` to fewer rows just for visualization, then scale back up for `tar_make()`.

**Reply by JohannesNE** [reply]
Good idea. It is not really a problem though. Static branching has
completely solved the problem and made my targets-setup much more
transparent.

**Comment by gorkang** [comment]
A bit late to the discussion, but I had the same need to have sensible names in dynamic branching. I ended up creating a target that renames the names of the dynamic output... but a `.names` parameter would be so much nicer :).

Just to elaborate a bit on my use case, I have a named list of dataframes, each one with a different Dependent Variable (used as the name for each DF of the list). The names can change, and I would like this to be used with other data, so dynamic branching seems a better fit.

In the first target I create a model and a table, and output them as a list.

In the renaming target, I use the `tar_meta(list_results)$children[[1]]` trick plus a `stri_replace_all_regex()` to rename the elements of the list. It takes a while and it's not pretty, The advantage is that then I can work with a list that contains elements that preserve meaning.

list_results$list_results_d68f7963_model becomes list_results_renamed$NAME1_model, where NAME1 is the name of the dependent variable of that analysis...

Anyway, just wanted to leave this here in case it is useful to someone, to check if there might be a better way, and to add visibility to this type of use of the dynamic branching.


```
# Functions -------------------------------------

analysis <-  function(DV_name, DF_analysis) {

  DF_model = DF_analysis[[DV_name]]
  model = lm(DV_name ~ Y, DF_model)
  table_model = create_table(model)

  return(list(model = model, table_model = table_model))

}

rename_targets <- function(DV_names, list_results) {

  NAMES = tibble(targets = tar_meta(list_results)$children[[1]], names = DV_names)
  names(list_results) = stringi::stri_replace_all_regex(names(list_results), pattern = NAMES$targets, replacement = NAMES$names, vectorize = FALSE)

  return(list_results)

}


# _targets -------------------------------------

tar_target(list_results, analysis(DV_name = names(DF_analysis) , DF = DF_analysis), pattern = map(names(DF_analysis))),

tar_target(list_results_renamed, rename_targets(DV_names, list_results)),


# RAW Outputs names -------------------

list_results$list_results_d68f7963_model
list_results$list_results_d68f7963_table_model
list_results$list_results_ea249feb_model
list_results$list_results_ea249feb_table_model

# Nicer output names --------------------

list_results_renamed$NAME1_model
list_results_renamed$NAME1_table_model
list_results_renamed$NAME2_model
list_results_renamed$NAME2_table_model
```
